### PR TITLE
Remove add another field button if it's dynamic zone from CTB

### LINKED ISSUE #19362: Component view with dynamic zone crashes app when trying to add more fields to said component
### Required System information

- Node.js version: 20
- Strapi version: 4.17.0 (still happening on `main` and `develop`)
- Is your project Javascript or Typescript: JS

### Describe the bug

When you add a dynamic zone to your schema in the CTB the UI is broken – the CTB thinks you want to add fields to the dynamic zone instead of the highlighted component within the CTB. If you press `add another field to this component` the app crashes because:

https://github.com/strapi/strapi/blob/a96121ddeb090c38e7553f9a520a4d97da965ebf/packages/core/content-type-builder/admin/src/components/FormModalHeader.tsx#L62

should have two optional chains because you can't index `undefined`. If you fix this, you then start adding fields directly to the DZ breaking schema validation. See image attached, there should be a list of fields for the selected component (a string called title).

### Steps to reproduce the behavior

1. Go to CTB
2. Make a component with just a title
3. make a content-type with a DZ field and add the aforementioned component
4. click "add another field to this component"
5. app will crash.

### Expected behavior

You should be able to see a list of the fields for each component listed in the DZ & then modify those components in situ of the DZ. Adding fields should only attach to the component & therefore validation of the schema should pass.

### Screenshots

CleanShot 2024-01-29 at 14 17 06@2x

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

Adds another check in the List component to not show "add another field" button incase the list was a dynamic zone list

### Why is it needed?

- Currently it's showing a button that shouldn't be there (Add another field button)
- When clicking the button an error occurs for the users

### How to test it?

How it currently is:
image

How it should look like:
image

How a component list should stay:
image

**Related Issues**
Fixes #19362

### CODE REVIEW

**Review by christiancp100 — APPROVED** [review]
Code LGTM

File: e2e/tests/content-type-builder/ctb.spec.ts
Comment (joshuaellis) [inline e2e/tests/content-type-builder/ctb.spec.ts:1]: can you name this better as to what you're actually testing please?

File: e2e/tests/content-type-builder/ctb.spec.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    await page.getByRole('link', { name: 'Shop' }).click();
+    await expect(
+      page.getByRole('button', {
+        name: 'Add another field to this component',
+      })
+    ).toBeVisible();
+    await expect(page.getByRole('button', { name: 'Add a component' })).toBeVisible();
+    await expect(
+      page.getByRole('button', { name: 'Add another field to this single type' })
+    ).toBeVisible();
```
Comment (joshuaellis) [inline e2e/tests/content-type-builder/ctb.spec.ts:30]: Can you look at the other e2e tests and see how most people have written comments to explain what's ging on?

it's not quite clear at the moment which might make it harder to maintain in the long run :)

**Review by joshuaellis — CHANGES_REQUESTED** [review]

File: e2e/tests/content-type-builder/ctb.spec.ts
Comment (joshuaellis) [reply e2e/tests/content-type-builder/ctb.spec.ts:1]: `editview` might be clearer?

**Review by joshuaellis — APPROVED** [review]
Great work!! ⭐

File: e2e/tests/content-type-builder/ctb-edit-view.spec.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+    // Navigate to the Shop single type and verify some essential buttons are visible
+    await page.getByRole('link', { name: 'Shop' }).click();
+    await expect(
+      page.getByRole('button', {
+        name: 'Add another field to this component',
+      })
+    ).toBeVisible();
+    await expect(page.getByRole('button', { name: 'Add a component' })).toBeVisible();
+    await expect(
```
Comment (innerdvations) [inline e2e/tests/content-type-builder/ctb-edit-view.spec.ts:31]: Glad you got the tests working! Could we add one that checks the button isn't visible on dzs?

I think you can use something like `expect(locator).toHaveCount(0)` on it, and to find it you can look for the button within the same row as "dynamic zone" so maybe something like `await expect(page.getByRole('row', { name: 'Dynamic Zone' }).getByRole('button', {name: 'Add a component'}).toHaveCount(0)`

**Review by innerdvations — APPROVED** [review]

**Pull request merged**